### ISSUE TITLE
servicebinding cannot be deleted if creation failed

### ISSUE THREAD

**Issue by gyliu513** [kubernetes-incubator/service-catalog#1623]
```
[root@gyliu-icp-1 2.jpress]# cat 2.2.jpress-binding.yaml
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: jpress-mariadb-binding
spec:
  instanceRef:
    name: jpress-mariadb-instance
  secretName: jpress-mariadb-credentials
```

```
[root@gyliu-icp-1 2.jpress]# kc get ServiceBinding -oyaml
apiVersion: v1
items:
- apiVersion: servicecatalog.k8s.io/v1beta1
  kind: ServiceBinding
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"servicecatalog.k8s.io/v1beta1","kind":"ServiceBinding","metadata":{"annotations":{},"name":"jpress-mariadb-binding","namespace":"default"},"spec":{"instanceRef":{"name":"jpress-mariadb-instance"},"secretName":"jpress-mariadb-credentials"}}
    creationTimestamp: 2017-12-16T02:30:39Z
    deletionGracePeriodSeconds: 0
    deletionTimestamp: 2017-12-16T02:43:32Z
    finalizers:
    - kubernetes-incubator/service-catalog
    generation: 2
    name: jpress-mariadb-binding
    namespace: default
    resourceVersion: "120953"
    selfLink: /apis/servicecatalog.k8s.io/v1beta1/namespaces/default/servicebindings/jpress-mariadb-binding
    uid: 1b3b7e54-e209-11e7-8725-fa163e4a45bf
  spec:
    externalID: 12660941-6a20-4a51-bbc1-b140631ff202
    instanceRef:
      name: jpress-mariadb-instance
    secretName: jpress-mariadb-credentials
  status:
    conditions:
    - lastTransitionTime: 2017-12-16T02:42:52Z
      message: 'Unbind call failed. Error unbinding from ServiceInstance "default/jpress-mariadb-instance"
        of ClusterServiceClass (K8S: "3533e2f0-6335-4a4e-9d15-d7c0b90b75b5" ExternalName:
        "mariadb") at ClusterServiceBroker "mariadb-broker": Status: 200; ErrorMessage:
        <nil>; Description: <nil>; ResponseError: unexpected end of JSON input'
      reason: UnbindCallFailed
      status: Unknown
      type: Ready
    - lastTransitionTime: 2017-12-16T02:42:52Z
      message: 'Unbind call failed. Error unbinding from ServiceInstance "default/jpress-mariadb-instance"
        of ClusterServiceClass (K8S: "3533e2f0-6335-4a4e-9d15-d7c0b90b75b5" ExternalName:
        "mariadb") at ClusterServiceBroker "mariadb-broker": Status: 200; ErrorMessage:
        <nil>; Description: <nil>; ResponseError: unexpected end of JSON input'
      reason: UnbindCallFailed
      status: "True"
      type: Failed
    orphanMitigationInProgress: false
    reconciledGeneration: 2
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

```
[root@gyliu-icp-1 2.jpress]# kc delete -f ./2.2.jpress-binding.yaml
servicebinding "jpress-mariadb-binding" deleted
```

The servicebinding was not deleted.
```
[root@gyliu-icp-1 2.jpress]# kc get servicebinding
NAME                     AGE
jpress-mariadb-binding   14m
```

**Comment by gyliu513** [comment]
Same issue also exist for `ServiceInstance`

**Comment by JacieChao** [comment]
I have met the same issue.

**Comment by singhwx** [comment]
I have met the same issue.

**Comment by JacieChao** [comment]
This issue has been fixed in v0.1.8
Duplicate of #1640

**Comment by mengjieli0726** [comment]
Same issue found in my env.
The sevicebinding can't be delete.
```
Events:
  Type     Reason              Age                From                                Message
  ----     ------              ----               ----                                -------
  Normal   InjectedBindResult  21m                service-catalog-controller-manager  Injected bind result
  Warning  UnbindCallFailed    16m (x2 over 16m)  service-catalog-controller-manager  Error unbinding from ServiceInstance "kube-system/wordpress-mariadb-instance" of ClusterServiceClass (K8S: "3533e2f0-6335-4a4e-9d15-d7c0b90b75b5" ExternalName: "mariadb") at ClusterServiceBroker "mariadb-broker": Status: 200; ErrorMessage: <nil>; Description: <nil>; ResponseError: unexpected end of JSON input
```
The serviceinstance delete depend servicebinding, in other words, servicebinding can't remove , which lead to serviceinstances can't be deleted.
```
Events:
  Type     Reason                                   Age                 From                                Message
  ----     ------                                   ----                ----                                -------
  Warning  ReferencesNonexistentServiceClass        47m (x48 over 57m)  service-catalog-controller-manager  References a non-existent ClusterServiceClass (ExternalName: "mariadb") or there is more than one (found: 0)
  Warning  DeprovisionBlockedByExistingCredentials  20s (x11 over 40m)  service-catalog-controller-manager  Delete instance blocked by existing ServiceBindings associated with this instance.  All credentials must be removed first
```

**Comment by jboyd01** [comment]
From the various comments on this issue it looks like each one is involving jpress-mariadb.   I'm guessing this is forked off from https://github.com/prydonius/mariadb-broker ?

The error is indicating the Broker returned a success (Status 200) on the unbind but the response body could not be unmarshaled.    https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#unbinding  indicates the body should be {}

It looks like the version of vendor/github.com/kubernetes-incubator/service-catalog/contrib/pkg/broker/server/server.go that you have has a bug:
```
func (s *server) unBind(w http.ResponseWriter, r *http.Request) {
        instanceID := mux.Vars(r)["instance_id"]
        bindingID := mux.Vars(r)["binding_id"]
        glog.Infof("UnBind: Service instance guid: %s:%s", bindingID, instanceID)

        if err := s.controller.UnBind(instanceID, bindingID); err == nil {
                w.WriteHeader(http.StatusOK)
                fmt.Print(w, "{}") //id)
        } else {
                util.WriteResponse(w, http.StatusBadRequest, err)
        }
}
```

Replace
```   fmt.Print(w, "{}") //id)```
with
```   fmt.Fprint(w, "{}") //id)```

**Comment by gyliu513** [comment]
Thanks @jboyd01 ，let me have a try.

**Comment by jboyd01** [comment]
@gyliu513 if possible, I recommend bumping up to the latest release of Service Catalog, we've had a lot of improvements since Catalog was vendored into your project.  That bug was fixed on our side about 9 months ago if I recall correctly.
